xdr: Fix uio reference for xdrmem ops.

Gaurav-Gangalwar opened the pull request:
We are not taking ref for uio in xdrmem_putbufs which causes leak as uio_release won't happen.

To fix memory leak reported by valigrind, while running IO on udp.

==5030== 441,058,920 (5,531,240 direct, 435,527,680 indirect) bytes in 53,185 blocks are definitely lost in loss record 439 of 439
==5030==    at 0x4C2C089: calloc (vg_replace_malloc.c:762)
==5030==    by 0x4F90773: xdr_READ3res_uio_setup (xdr_nfs23.c:825)
==5030==    by 0x4F908EE: xdr_READ3resok_encode (xdr_nfs23.c:854)
==5030==    by 0x4F909DF: xdr_READ3resok (xdr_nfs23.c:873)
==5030==    by 0x4F90AA7: xdr_READ3res (xdr_nfs23.c:895)
==5030==    by 0x52AECC5: svcauth_none_wrap (svc_auth_none.c:45)
==5030==    by 0x52B023D: svc_dg_reply (svc_dg.c:423)
==5030==    by 0x52AE231: svc_sendreply (svc.c:506)
==5030==    by 0x4E91A7E: complete_request (nfs_worker_thread.c:784)
==5030==    by 0x4E94093: nfs_rpc_process_request (nfs_worker_thread.c:1520)
==5030==    by 0x4E94669: nfs_rpc_valid_NFS (nfs_worker_thread.c:1724)
==5030==    by 0x52AFFCC: svc_dg_decode (svc_dg.c:362)

Signed-off-by: Gaurav Gangalwar <gaurav.gangalwar@nutanix.com>

Gaurav-Gangalwar commented:
@ffilz

ffilz commented on src/xdr_mem.c:
```diff
@@ -183,6 +183,8 @@ xdrmem_putbufs(XDR *xdrs, xdr_uio *uio, u_int flags)
 	for (ix = 0; ix < uio->uio_count; ++ix) {
 		xdr_vio *v = &(uio->uio_vio[ix]);
 
+		uio->uio_references++;
```
This doesn't look right, this is in a loop...

Also would help to have more explanation as to why we are replacing a decrement of refcount with an increment (in a loop no less). Where is the matching decrement(s)?

Gaurav-Gangalwar replied:
looks like we are missing to take uio_reference in xdrmem_putbufs

uio_references is 0 when we allocated uio in xdr_READ3res_uio_setup
```
(gdb) 
825       uio = gsh_calloc(1, sizeof(struct xdr_uio) + sizeof(struct xdr_vio));
(gdb) 
826       uio->uio_release = xdr_READ3res_uio_release;
(gdb) 
827       uio->uio_count = 1;
(gdb) 
828       uio->uio_vio[0].vio_base = objp->data.data_val;
(gdb) 
829       uio->uio_vio[0].vio_head = objp->data.data_val;
(gdb) 
830       uio->uio_vio[0].vio_tail = objp->data.data_val + size2;
(gdb) 
831       uio->uio_vio[0].vio_wrap = objp->data.data_val + size2;
(gdb) 
832       uio->uio_vio[0].vio_length = size2;
(gdb) p *uio
$1 = {uio_refer = 0x0, uio_release = 0x7f47c698261d <xdr_READ3res_uio_release>, uio_p1 = 0x0, uio_p2 = 0x0, uio_u1 = 0x0, uio_u2 = 0x0, uio_count = 1, 
  uio_flags = 0, uio_references = 0, uio_vio = 0x7f47ba48f0c0}
(gdb) p uio
$2 = (struct xdr_uio *) 0x7f47ba48f080
```

Now in xdrmem_putbufs we decrement and check uio_references == 0 to release uio, which fails since we never took the ref and after decrementing uio_references = -1
```
Breakpoint 2, xdrmem_putbufs (xdrs=0x7f47ba4191c8, uio=0x7f47ba48f080, flags=0) at /home/nutanix/nfs-ganesha/src/libntirpc/src/xdr_mem.c:177
177       xdr_tail_update(xdrs);
(gdb)
179       __warnx(TIRPC_DEBUG_FLAG_XDR,
(gdb)
183       for (ix = 0; ix < uio->uio_count; ++ix) {
(gdb) p uio
$3 = (xdr_uio *) 0x7f47ba48f080
(gdb) p *uio
$4 = {uio_refer = 0x0, uio_release = 0x7f47c698261d <xdr_READ3res_uio_release>, uio_p1 = 0x0, uio_p2 = 0x0, uio_u1 = 0x0, uio_u2 = 0x0, uio_count = 1,
  uio_flags = 0, uio_references = 0, uio_vio = 0x7f47ba48f0c0}
(gdb) n
184        xdr_vio *v = &(uio->uio_vio[ix]);
(gdb)
186        if (!XDR_PUTBYTES(xdrs, v->vio_head, v->vio_length))
(gdb)
189        __warnx(TIRPC_DEBUG_FLAG_XDR,
(gdb)
183       for (ix = 0; ix < uio->uio_count; ++ix) {
(gdb)
196       if (!(--uio->uio_references)) {
(gdb)
207       return (TRUE);
(gdb) p uio
$5 = (xdr_uio *) 0x7f47ba48f080
(gdb) p *uio
$6 = {uio_refer = 0x0, uio_release = 0x7f47c698261d <xdr_READ3res_uio_release>, uio_p1 = 0x0, uio_p2 = 0x0, uio_u1 = 0x0, uio_u2 = 0x0, uio_count = 1,
  uio_flags = 0, uio_references = -1, uio_vio = 0x7f47ba48f0c0}
(gdb) 
```

So I just added the fix to take ref by incrementing uio_references, we always keep uio_count=1, so uio_references should be 1, xdr_READ3res_uio_release will decrement the ref taken.

```
        if (!(--uio->uio_references)) {
                for (ix = 0; ix < uio->uio_count; ix++) {
                        gsh_free(uio->uio_vio[ix].vio_base);
                }
                gsh_free(uio);
        }
```

we follow same in xdr_ioq_putbufs, where we increment in loop but decrement once.

ffilz replied:
While currently, uio_count is always 1, it could become > 1 if we ever allow vector read/write. I'd rather make sure the code works OK if uio_count is ever > 1.

Gaurav-Gangalwar replied:
I think we use xdr_ioq_uv for vector IO.
Also I added assert for assert((uio->uio_references == 1)) so that we won't proceed if uio_count > 1

Gaurav-Gangalwar committed:
Fixes to enabled NFS over RDMA
1>Connection management.
2>Scalable per connection buffer management to use/reuse rdma registered
buffers for NFS request/responses/data buffers.
3>Fixes to make rdma_read and rdma_write work.
4>Fixes to make reply_list and write_list work.
5>Limit rdma connections.

We have supporting patch in Ganesha to use rdma registered for read and
readdir responses.
With this patch and supporting Ganesha patch, we should be able to use
NFSv3 and NFSv4.0 over RDMA.
Delegation has to be disabled for NFSv4.0, since callback channel is not
yet supported over RDMA.
Connection disconnect or umount needs some more fixes to do cleanup,
we will fix it as seperate patch.

ffilz commented:
It looks like you pushed the RDMA changes on top of this other patch.

ffilz commented:
Could you submit a new pull request, for some reason, your RDMA changes are showing up in an xdrmem ops pull request

Gaurav-Gangalwar commented:
let me close this one
